### PR TITLE
Adds auto-population to the cookbook

### PR DESCRIPTION
**Pull request by nmajask**
## About the PR

Adds auto-population for food items into the cookbook, as well as produce in the botany guidebook and medical items with microwave recipes in the chemistry guidebook. This is my first time doing something big in C# and working with the UI code, so the code quality is very low and if I could get feedback on it that would be amazing.

The food item embed contain an image of the food, the food's description, all of the recipes that result in it, and all of the recipes that require it. The only recipes that work are microwave recipes and reactions that spawn entities.

I'm not the best when it comes to knowing about food, so please give feedback on the categorization.

## Why / Balance

Currently, the only way to learn recipes is by looking it up in the wiki which is outdated or going through the code which can be annoying to do. This makes it easier for chefs to know what they can even cook, and also shows them what things can be used for.

## Technical details

So a lot of this has low code quality as I'm new to C# and xaml. One of the main issues is the fact that reaction effects which is what spawns items are held server side, so I had to graft another list to the system that handles chem effects. Information being stored server side is one of the main issues i have encountered. I would like to combine foods with their sliced versions, but the slice-able food comp is held server side. I want to add reagent grinding to the guidebook for chem, but the info for that is held server side. I want to add what the foods taste like to the description, but that is held server side. We would probably just need a generic system to get information for the guidebook to the client but that is beyond me at the moment.

As for the UI, I made some changes to avoid redundant code. Reagent recipes have been split of into a new reaction embed to allow for me to use them for food item recipes. Reactions also now show solid products and show units when the reaction is quantized to show that you need exact amounts. Microwave recipes are also a new embed type similar to the reaction embed, and they show the time needed for it to cook. For the food embed, it doesn't have to be given a food to embed, it can be given any entity. It automatically lists any reactions or microwave recipes that produce it, and any microwave recipes that require it. There is also a cookbook group embed that lists all the foods of a specific category similar to the reagent groups. To add an entity to the group, you need to give it a CookbookDocumentation comp and a catagory.

I've tried to avoid messing with the reagent embeds where possible to avoid bloat in the PR, though there are a couple things that kind of creep out of the scope of the PR like removing the extra unused bun prototype and making some of the formatting of some of the yml documents more consistent. I also haven't done the best job of splitting stuff into different commits because I was still learning.

## Media

image
The embed

image
This is what hell looks like
- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: The guidebook for cooking recipes now automatically updates and has a new format

### CODE REVIEW

**Comment by deltanedas** [comment]
i dont think Or is needed since theyre separate recipes, not multiple recipes for the same thing

**Comment by nmajask** [comment]
> i dont think Or is needed since theyre separate recipes, not multiple recipes for the same thing

Its mostly to visually separate the recipes, but it does do a bad job at that anyways. I would use the list thing that the crew monitor uses but I couldnt get that to work.

**Comment by nmajask** [comment]
PR is ready for review, just need feedback on the implementation and where to categorize foods.

**Comment by daerSeebaer** [comment]
Your current version is failing in the integration tests. I don't know exactly what they do, but the error was thrown in a file called "GuideEntryPrototypeTests.cs" so I guess that is an error related to this PR. Please look into this and maybe ask someone with more knowledge about the test pipeline what the Error means.

**Review by daerSeebaer — CHANGES_REQUESTED** [review]

File: .vscode/settings.json [hunk 1]
```diff
@@ -1,4 +1,5 @@
 {
     "omnisharp.analyzeOpenDocumentsOnly": true,
-    "dotnet.defaultSolution": "SpaceStation14.sln"
+    "dotnet.defaultSolution": "SpaceStation14.sln",
+    "git-blame.gitWebUrl": ""
```
[the remark this answers is not in the document]
Comment (nmajask) [reply .vscode/settings.json:4]: Im using VSC to code and it constantly makes this change, forgot to not push it in one of the commits

File: Content.Client/Guidebook/Controls/GuideCookbookGroupEmbed.xaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+﻿<BoxContainer xmlns="https://spacestation14.io"
+              Orientation="Vertical">
+    <BoxContainer Name="CatagoryContainer" Orientation="Vertical"/>
```
[the remark this answers is not in the document]
Comment (nmajask) [reply Content.Client/Guidebook/Controls/GuideCookbookGroupEmbed.xaml]: I cant spell for shit, thanks for pointing that out. Should be fixed in the latest commit.

File: Content.Client/Guidebook/Controls/GuideMicrowaveRecipeEmbed.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        var productPrototype = _prototype.Index<EntityPrototype>(recipePrototype.Result);
+        ProductIcon.Textures = SpriteComponent.GetPrototypeTextures(productPrototype, _resources).Select(o => o.Default).ToList();
+        var productMsg = new FormattedMessage();
+        productMsg.AddMarkup(Loc.GetString("guidebook-food-microwave-entity-display",
+            ("name", productPrototype.Name),
+            ("amount", 1)));
+        productMsg.Pop();
+        ProductLabel.SetMessage(productMsg);
+    }
+}
```
Comment (metalgearsloth) [inline Content.Client/Guidebook/Controls/GuideMicrowaveRecipeEmbed.xaml.cs:133]: Please document your code.

File: Content.Client/Guidebook/Controls/GuideReactionEmbed.xaml.cs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+                productMsg.AddMarkup(Loc.GetString("guidebook-reaction-prototype-display",
+                    ("name", _prototype.Index<EntityPrototype>(product).Name), ("ratio", amount)));
+                u++;
+                if (u < productCount)
+                    productMsg.PushNewline();
+            }
+        }
+        productMsg.Pop();
+        ProductsLabel.SetMessage(productMsg);
+    }
```
Comment (metalgearsloth) [inline Content.Client/Guidebook/Controls/GuideReactionEmbed.xaml.cs:142]: As above "u" is not a clear variable name and the code is unnecessarily hard to read.

File: Content.Server/Chemistry/EntitySystems/ChemistryGuideDataSystem.cs [hunk 5]
```diff
@@ -26,12 +29,26 @@ public override void Initialize()
 
     private void InitializeServerRegistry()
     {
-        var changeset = new ReagentGuideChangeset(new Dictionary<string, ReagentGuideEntry>(), new HashSet<string>());
+        var changeset = new ReagentGuideChangeset(new Dictionary<string, ReagentGuideEntry>(), new HashSet<string>(), new Dictionary<string, Dictionary<string, uint>>(), new HashSet<string>());
```
Comment (metalgearsloth) [inline Content.Server/Chemistry/EntitySystems/ChemistryGuideDataSystem.cs:32]: If these are just being initialized to default values there's no point having them in the ctor.

File: Content.Server/Chemistry/EntitySystems/ChemistryGuideDataSystem.cs [hunk 6]
```diff
@@ -26,12 +29,26 @@ public override void Initialize()
...
+        {
+            foreach (CreateEntityReactionEffect effect in proto.Effects.OfType<CreateEntityReactionEffect>())
+            {
+                if (!changeset.ReactionSolidProductEntries.ContainsKey(proto.ID))
+                {
+                    changeset.ReactionSolidProductEntries.Add(proto.ID, new Dictionary<string, uint>());
+                    ReactionRegistry.Add(proto.ID, new Dictionary<string, uint>());
+                }
+                changeset.ReactionSolidProductEntries[proto.ID].Add(effect.Entity, effect.Number);
+                ReactionRegistry[proto.ID].Add(effect.Entity, effect.Number);
```
Comment (metalgearsloth) [inline Content.Server/Chemistry/EntitySystems/ChemistryGuideDataSystem.cs:50]: Use GetOrNew.

File: Content.Server/Chemistry/EntitySystems/ChemistryGuideDataSystem.cs [hunk 7]
```diff
@@ -42,8 +59,7 @@ private void OnPlayerStatusChanged(object? sender, SessionStatusEventArgs e)
     {
         if (e.NewStatus != SessionStatus.Connected)
             return;
-
-        var sendEv = new ReagentGuideRegistryChangedEvent(new ReagentGuideChangeset(Registry, new HashSet<string>()));
+        var sendEv = new ReagentGuideRegistryChangedEvent(new ReagentGuideChangeset(ReagentRegistry, new HashSet<string>(), ReactionRegistry, new HashSet<string>()));
```
Comment (metalgearsloth) [inline Content.Server/Chemistry/EntitySystems/ChemistryGuideDataSystem.cs:62]: As above.

File: Content.Server/Chemistry/EntitySystems/ChemistryGuideDataSystem.cs [hunk 8]
```diff
@@ -52,14 +68,14 @@ private void PrototypeManagerReload(PrototypesReloadedEventArgs obj)
         if (!obj.ByType.TryGetValue(typeof(ReagentPrototype), out var reagents))
             return;
 
-        var changeset = new ReagentGuideChangeset(new Dictionary<string, ReagentGuideEntry>(), new HashSet<string>());
+        var changeset = new ReagentGuideChangeset(new Dictionary<string, ReagentGuideEntry>(), new HashSet<string>(), new Dictionary<string, Dictionary<string, uint>>(), new HashSet<string>());
```
Comment (metalgearsloth) [inline Content.Server/Chemistry/EntitySystems/ChemistryGuideDataSystem.cs:72]: As above.

File: Content.Shared/Kitchen/Components/CookbookDocumentationComponent.cs [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+public sealed partial class CookbookDocumentationComponent : Component
+{
+    /// <summary>
+    /// What guides to include show when opening the guidebook. The first entry will be used to select the currently
+    /// selected guidebook.
+    /// </summary>
+    [DataField("category")]
+    [ViewVariables]
+    public string Category = "Unknown";
+}
```
Comment (metalgearsloth) [inline Content.Shared/Kitchen/Components/CookbookDocumentationComponent.cs:18]: Category needs to be a prototype so it gets properly validated, datafield strings are obsolete, and this needs networking + access attribute.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by metalgearsloth** [comment]
Re-open if you come back to it.

**Pull request closed without merging**